adaptdl/adaptdl/torch/scaling_rules.py

```diff
@@ -282,49 +416,24 @@ def _final_callback(self):
             theta = self._smoothing ** scale
             self._update_avg('norm_avg', grad_sqr, theta)
             self._update_avg('var_avg', grad_var, theta)
-        self._local_sqr = None
+        self._local_vars.reset()
 
-    def step(self, *args, **kwargs):
-        """
-        Run one optimizer step using Adascale. Essentially just invokes
-        ``optimizer.step(*args, **kwargs)`` with a scaled learning rate.
+    def _calculate_scale_multiplier(self, pg_idx=None) -> float:
+        if pg_idx is None:
+            raise ValueError("Parameter index must be provided.")
+        grad_sqr = float(self._state["norm_avg"][pg_idx])
+        grad_var = float(self._state["var_avg"][pg_idx])
+        gain = (grad_var + grad_sqr) / (grad_var / self._scale + grad_sqr)
+        return gain
 
-        Arguments:
-            args: Positional arguments passed to ``optimizer.step``.
-            kwargs: Keyword arguments passed to ``optimizer.step``.
-        """
-        if (self.is_accumulation_step()):
-            self._current_accumulation_step += 1
-            self._made_step = False
-            return
 
-        self._current_accumulation_step = 0
-        self._made_step = True
-        initial_lr = [pg["lr"] for pg in self._optimizer.param_groups]
-        for idx, param_group in enumerate(self._optimizer.param_groups):
-            grad_sqr = float(self._state["norm_avg"][idx])
-            grad_var = float(self._state["var_avg"][idx])
-            gain = (grad_var + grad_sqr) / (grad_var / self._scale + grad_sqr)
-            param_group["lr"] = gain * param_group["lr"]
-        self._optimizer_step(*args, **kwargs)
-        for lr, param_group in zip(initial_lr, self._optimizer.param_groups):
-            param_group["lr"] = lr
+class LinearScale(ScalingRuleBase):
```

Comment:
`get_progress` will need to be added to these classes too. I think having it return scale for this one should be fine (and sqrt(scale) for the `SqrtScale` class).

Reply:
I don't think this is correct mathematically. Progress should be calculated the same way as AdaScale in all cases. Maybe factor out the norm/var estimation code and the progress calculation code out of the scaling rule classes.